### ISSUE TITLE
Grid multiselection is very slow

### ISSUE THREAD

**Issue by Slugger2k** [vaadin/vaadin-grid-flow#451]
When i create a grid with about 4k entries, the multiselection is very slow. Also the scrolling of a full selected grid is slow.

**Comment by robeden** [comment]
This is also an issue in V8 and you can duplicate it by clicking the "select all" check. Doing this with a Grid containing 250k rows will lock up the session for 15+ minutes.

**Comment by Wnt** [comment]
Even with just 400 items in a Grid makes the client-side spend > 10 seconds in  `grid.$connector.doSelection`

![image](https://user-images.githubusercontent.com/886091/54214108-4b6f4780-44ee-11e9-902c-9c20062ef50a.png)

Here's minimal Java code to replicate the issue:

```java
package org.vaadin.jonni;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.UUID;
import com.vaadin.flow.component.grid.Grid;
import com.vaadin.flow.component.grid.Grid.SelectionMode;
import com.vaadin.flow.component.orderedlayout.VerticalLayout;
import com.vaadin.flow.router.Route;

@Route("")
public class MainView extends VerticalLayout {

	public MainView() {
		Grid<Dto> grid = new Grid<>(Dto.class);
		List<Dto> data = new ArrayList<>();
		for (int i = 0; i < 400; i++) {
			data.add(getNewMockDto());
		}
		grid.setItems(data);
		grid.setSelectionMode(SelectionMode.MULTI);
		grid.getColumns().forEach(col -> {
			col.setResizable(true);
			col.setWidth("100px");
		});
		add(grid);
		expand(grid);
		setSizeFull();
	}

	private Dto getNewMockDto() {
		return new Dto(UUID.randomUUID().toString());
	}

	public static class Dto {

		private String property1;

		public Dto(String property1) {
			super();
			this.property1 = property1;
		}

		public String getProperty1() {
			return property1;
		}

		public void setProperty1(String property1) {
			this.property1 = property1;
		}

	}
}

```

Using Vaadin 13.0.1, Windows 10, Chrome 72.

**Comment by gsedlacz** [comment]
Hello,
do you have rough plan when this will be fixed ?
Thank you in advance.

**Comment by mstahv** [comment]
I think there are two performance issues here. The old and tradition, the one on server-side and on Vaadin 8 too that forces basically all entitties to be loaded, even with lazy loading, and then the other, which is in the current web component/Polymer based Grid, tracked down by @Wnt. I guess the client-side performance issue should be in vaadin-grid repository instead. 

I wonder who would be the best expert to look into this? @web-padawan or @tomivirkki ? Somebody else?

**Comment by kolle1986** [comment]
Is this resolved? Still happening on vaadin 14 grid

**Comment by mimarko** [comment]
Even with list of 1k rows select all works very slow, any workaround in meantime ?

**Comment by OlliTietavainenVaadin** [comment]
Sample project: [demov14-grid.zip](https://github.com/vaadin/vaadin-grid-flow/files/3381699/demov14-grid.zip)

**Comment by pekam** [comment]
> I think there are two performance issues here. The old and tradition, the one on server-side and on Vaadin 8 too that forces basically all entitties to be loaded, even with lazy loading, and then the other, which is in the current web component/Polymer based Grid, tracked down by Wnt. I guess the client-side performance issue should be in vaadin-grid repository instead.

Even the client-side performance issue seems to be only related to the Flow component. The function which takes a lot of time in the screenshot, `grid.$connector.doSelection`, is actually part of the Flow component implementation. The Flow component also has its own custom implementation of the selection column, i.e. it's not using the web component's selection column element.

I also ran a quick test with the [web component's selection demo](https://cdn.vaadin.com/vaadin-grid/5.4.6/demo/#grid-selection-demos) and Chrome dev tools. Setting 100,000 items to the grid and then selecting all took only 0.x seconds.

When the web component has a lazy data provider instead of items-array, it doesn't even show the select-all checkbox by default. The Flow component on the other hand provides this checkbox, which loads the entire backend into server memory so it can return all the selected items with e.g. `Grid::getSelectedItems`.

But for the client-side execution, there seems to be at least some things we can do to improve the performance:
1. `$connector.doSelection` is called also for those items that are not even loaded. It seems that the item-keys for selected items are stored already at this time at the client, although those are probably not needed until the items are actually loaded from the server.
2. Currently Flow registers a separate JavaScript call to `$connector.doSelection` per each item, which includes some overhead compared to sending a single JavaScript execution.

**Comment by pleku** [comment]
By default the select all checkbox should not be even visible for any lazy loading data provider grid.
For in memory data providers, the checkbox is there by default but can be explicitly made hidden from the selection model https://github.com/vaadin/vaadin-grid-flow/blob/master/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/GridMultiSelectionModel.java#L112

The discovered performance issues / bugs on the client side connector should be fixed.
For the server side
1) having a lazy loading data provider and select all doesn't make sense. There are just some methods that should not be called in this case (like get all selected items) but instead there should be API for checking if all is selected.
2) for in memory data providers and select all, the select all state could be handled better than currently. I suspect there might be some low hanging fruits to pick for this. Similarly  as in the lazy case, the special state of selecting all should not require any excess bookkeeping for all items, or when the user starts deselecting items, it would be easier to update just the deselected items and not reselect all still selected. Though just a hunch.

Anyways, one fix at a time should be a good way to start improving this.

**Comment by pekam** [comment]
Vaadin 14.0.0.rc7 includes the first performance fix #715. It fixes the inefficient client-side processing that can be seen in @Wnt's screenshot.

Another fix #723, which makes the message sizes and client-side processing times constant, instead of relative to the full size of the data set, is still in progress.

See PR descriptions for more details and performance measurements.

**Comment by pekam** [comment]
With #715 and #723, I'd consider this problem solved, so I'm closing the issue.

The performance could be further optimized, but this ticket mostly describes multiselection being unusable even with reasonable amount of items, which is not the case anymore.

The second fix #723 will be included in `vaadin-grid-flow` 4.0.4 which will be probably included in Vaadin 14.0.0.rc8
